File: iconanim/src/main/java/sgtmelon/iconanim/library/SwitchButton.java
package sgtmelon.iconanim.library;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;

import androidx.annotation.CallSuper;
import androidx.annotation.ColorInt;
import androidx.annotation.DrawableRes;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.core.content.ContextCompat;
import sgtmelon.iconanim.R;
import sgtmelon.iconanim.office.intf.AnimIntf;

/**
 * Кнопка с автоматизацией процесса смены иконки
 */
public class SwitchButton extends AppCompatImageButton implements AnimIntf {

    public final static int SRC_NULL = -1;

    final Context context;

    @DrawableRes protected int srcDisable, srcSelect;
    @DrawableRes protected int srcDisableAnim, srcSelectAnim;
    @ColorInt protected int srcDisableColor, srcSelectColor;

    @Nullable protected Drawable drawableDisable;
    @Nullable protected Drawable drawableSelect;


    public SwitchButton(Context context, AttributeSet attrs) {
        super(context, attrs);

        this.context = context;

        setupAttribute(attrs);
        setupDrawable();
    }

    public SwitchButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        this.context = context;

        setupAttribute(attrs);
        setupDrawable();
    }

    public final void setSrcDisable(@DrawableRes int srcDisable) {
        this.srcDisable = srcDisable;
    }

    public final void setSrcSelect(@DrawableRes int srcSelect) {
        this.srcSelect = srcSelect;
    }

    public final void setSrcDisableAnim(@DrawableRes int srcDisableAnim) {
        this.srcDisableAnim = srcDisableAnim;
    }

    public final void setSrcSelectAnim(@DrawableRes int srcSelectAnim) {
        this.srcSelectAnim = srcSelectAnim;
    }

    public final void setSrcDisableColor(@ColorInt int srcDisableColor) {
        this.srcDisableColor = srcDisableColor;
    }

    public final void setSrcSelectColor(@ColorInt int srcSelectColor) {
        this.srcSelectColor = srcSelectColor;
    }

    private void setupAttribute(AttributeSet attrs) {
        final TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.SwitchButton);

        srcDisable = array.getResourceId(R.styleable.SwitchButton_srcDisable, SRC_NULL);
        srcSelect = array.getResourceId(R.styleable.SwitchButton_srcSelect, SRC_NULL);

        srcDisableAnim = array.getResourceId(R.styleable.SwitchButton_srcDisableAnim, SRC_NULL);
        srcSelectAnim = array.getResourceId(R.styleable.SwitchButton_srcSelectAnim, SRC_NULL);

        srcDisableColor = array.getColor(R.styleable.SwitchButton_srcDisableColor, Color.BLACK);
        srcSelectColor = array.getColor(R.styleable.SwitchButton_srcSelectColor, Color.BLACK);

        array.recycle();
    }

    @CallSuper
    public void setupDrawable() {
        if (srcDisable != SRC_NULL) {
            drawableDisable = ContextCompat.getDrawable(context, srcDisable);
            if (drawableDisable != null) {
                drawableDisable.setColorFilter(srcDisableColor, PorterDuff.Mode.SRC_ATOP);
            }
        }

        if (srcSelect != SRC_NULL) {
            drawableSelect = ContextCompat.getDrawable(context, srcSelect);
            if (drawableSelect != null) {
                drawableSelect.setColorFilter(srcSelectColor, PorterDuff.Mode.SRC_ATOP);
            }
        }

        setImageDrawable(drawableSelect);
    }

    @Override
    public void setDrawable(boolean select, boolean needAnim) {
        setImageDrawable(select
                ? drawableSelect
                : drawableDisable);
    }

}

File: iconanim/src/main/java/sgtmelon/iconanim/widget/SwitchButton.kt
package sgtmelon.iconanim.widget

import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.AnimatedVectorDrawable
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.appcompat.widget.AppCompatImageButton
import androidx.core.graphics.BlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import sgtmelon.common.utils.getDrawableCompat
import sgtmelon.common.utils.setColorFilterCompat
import sgtmelon.iconanim.IconAnimControl
import sgtmelon.iconanim.IconBlockCallback
import sgtmelon.iconanim.IconChangeCallback
import sgtmelon.iconanim.R
import sgtmelon.iconanim.control.IIconAnimControl

/**
 * Button with automatic icon change via [setDrawable] func.
 */
class SwitchButton(
    context: Context,
    attrs: AttributeSet
) : AppCompatImageButton(context, attrs),
    IconChangeCallback {

    @DrawableRes private val srcDisable: Int
    @DrawableRes private val srcSelect: Int
    @DrawableRes private val srcDisableAnim: Int
    @DrawableRes private val srcSelectAnim: Int

    @ColorInt private val srcDisableColor: Int
    @ColorInt private val srcSelectColor: Int

    private val iconDisable: Drawable?
    private val iconSelect: Drawable?
    private val iconDisableAnim: AnimatedVectorDrawable?
    private val iconSelectAnim: AnimatedVectorDrawable?

    private val animControl: IIconAnimControl

    init {
        val array = context.obtainStyledAttributes(attrs, R.styleable.SwitchButton)

        srcDisable = array.getResourceId(R.styleable.SwitchButton_srcDisable, ND_SRC)
        srcSelect = array.getResourceId(R.styleable.SwitchButton_srcSelect, ND_SRC)

        srcDisableAnim = array.getResourceId(R.styleable.SwitchButton_srcDisableAnim, ND_SRC)
        srcSelectAnim = array.getResourceId(R.styleable.SwitchButton_srcSelectAnim, ND_SRC)

        srcDisableColor = array.getColor(R.styleable.SwitchButton_srcDisableColor, Color.BLACK)
        srcSelectColor = array.getColor(R.styleable.SwitchButton_srcSelectColor, Color.BLACK)

        array.recycle()

        iconDisable = if (srcDisable != ND_SRC) context.getDrawableCompat(srcDisable) else null
        iconSelect = if (srcSelect != ND_SRC) context.getDrawableCompat(srcSelect) else null

        iconDisableAnim = if (srcDisableAnim != ND_SRC) {
            context.getDrawableCompat(srcDisableAnim) as? AnimatedVectorDrawable
        } else {
            null
        }
        iconSelectAnim = if (srcSelectAnim != ND_SRC) {
            context.getDrawableCompat(srcSelectAnim) as? AnimatedVectorDrawable
        } else {
            null
        }

        animControl = IconAnimControl(context, iconSelectAnim, iconDisableAnim, changeCallback = this)
    }

    init {
        iconDisable?.setColorFilterCompat(srcDisableColor)
        iconSelect?.setColorFilterCompat(srcSelectColor)
        iconDisableAnim?.setColorFilterCompat(srcDisableColor)
        iconSelectAnim?.setColorFilterCompat(srcSelectColor)
    }

    fun setBlockCallback(blockCallback: IconBlockCallback) {
        animControl.blockCallback = blockCallback
    }

    override fun setDrawable(isEnterIcon: Boolean, needAnim: Boolean) {
        if (needAnim) {
            setImageDrawable(animControl.getIcon(isEnterIcon))
        } else {
            setImageDrawable(if (isEnterIcon) iconSelect else iconDisable)
        }
    }

    companion object {
        const val ND_SRC = -1
    }
}